File: Problems/DesignUndergroundSystem.py
class UndergroundSystem:

    def __init__(self):
        self.checkInInformation = {}
        self.timeTotal = {}

    def checkIn(self, id: int, stationName: str, t: int) -> None:
        self.checkInInformation[id] = (stationName, t)

    def checkOut(self, id: int, stationName: str, t: int) -> None:
        startStation, startTime = self.checkInInformation[id]
        if (startStation, stationName) in self.timeTotal:
            self.timeTotal[(startStation, stationName)].append(t - startTime)
        else:
            self.timeTotal[(startStation, stationName)] = [t - startTime]

    def getAverageTime(self, startStation: str, endStation: str) -> float:
        return float(sum(self.timeTotal[(startStation, endStation)])) / len(self.timeTotal[(startStation, endStation)])


# Your UndergroundSystem object will be instantiated and called as such:
# obj = UndergroundSystem()
# obj.checkIn(id,stationName,t)
# obj.checkOut(id,stationName,t)
# param_3 = obj.getAverageTime(startStation,endStation)

File: JSProblems/DesignUndergroundSystem.js
var UndergroundSystem = function() {
    this.average = new Map();
    this.train = new Map();    
};

/** 
 * @param {number} id 
 * @param {string} stationName 
 * @param {number} t
 * @return {void}
 */
UndergroundSystem.prototype.checkIn = function(id, stationName, t) {
    this.train.set(id, [stationName, t]);    
};

/** 
 * @param {number} id 
 * @param {string} stationName 
 * @param {number} t
 * @return {void}
 */
UndergroundSystem.prototype.checkOut = function(id, stationName, t) {
    const [start, s] = this.train.get(id);
    const key = [start, stationName].join();
    if(this.average.has(key)){
        let [average, count] = this.average.get(key);
        this.average.set(key, [average * (count/++count) + ((t-s)/count), count]);
    }
    else{
        this.average.set(key, [(t-s),1]);
    }
    this.train.delete(id);
};

/** 
 * @param {string} startStation 
 * @param {string} endStation
 * @return {number}
 */
UndergroundSystem.prototype.getAverageTime = function(startStation, endStation) {
    return this.average.get([startStation, endStation].join())[0];
};

/** 
 * Your UndergroundSystem object will be instantiated and called as such:
 * var obj = new UndergroundSystem()
 * obj.checkIn(id,stationName,t)
 * obj.checkOut(id,stationName,t)
 * var param_3 = obj.getAverageTime(startStation,endStation)
 */
